scripts/pylib/twister/twisterlib/runner.py

```diff
@@ -1979,74 +1973,119 @@ def add_tasks_to_queue(
+            self, instance: TestInstance, processing_ready: dict[str, TestInstance]
+    ):
+        # Verify that all required applications were successfully built
+        found_failed_app = False
+        for required_app in instance.required_applications:
+            inst = processing_ready.get(required_app)
+            if inst.status not in (TwisterStatus.PASS, TwisterStatus.NOTRUN):
+                logger.debug(f"{required_app}: Required application failed: {inst.reason}")
+                found_failed_app = True
+        return found_failed_app
+
+    def required_applications_processed(self, instance: TestInstance, processing_queue: deque,
+                                        processing_ready: dict[str, TestInstance], task):
+        if not instance.required_applications:
+            return True
+
+        if not self._required_applications_are_ready(instance, processing_ready):
+            # required app not ready yet,
+            # add the task back to the pipeline to process it later
+            if self.jobs > 1:
+                # to avoid busy waiting
+                time.sleep(1)
+            processing_queue.appendleft(task)
+            return False
+
+        if self._required_applications_failed(instance, processing_ready):
+            instance.status = TwisterStatus.SKIP
+            for tc in instance.testcases:
+                tc.status = TwisterStatus.SKIP
+            instance.reason = "Required application failed"
+            instance.required_applications = []
+            processing_queue.append({"op": "report", "test": instance})
+            return False
+
+        # keep paths to required applications build directories to use it in harness module
+        for required_image in instance.required_applications:
+            instance.required_build_dirs.append(self.instances[required_image].build_dir)
+
+        # required applications are ready, clear to not process them later
+        instance.required_applications = []
+        return True
+
+    def _pipeline_mgr(self, processing_queue: deque, processing_ready: dict[str, TestInstance],
+                      lock, results: ExecutionCounter):
+        while True:
+            try:
+                task = processing_queue.pop()
+            except IndexError:
+                break
+            else:
+                instance: TestInstance = task['test']
 
-    def pipeline_mgr(self, pipeline, done_queue, lock, results):
+                if not self.required_applications_processed(
+                    instance, processing_queue, processing_ready, task
+                ):
+                    # postpone processing task if required applications are not ready
+                    continue
+
+                pb = ProjectBuilder(instance, self.env, self.jobserver)
+                pb.duts = self.duts
+                pb.process(processing_queue, processing_ready, task, lock, results)
+                if self.env.options.quit_on_failure and \
```

Comment:
nit: Use parentheses instead of `\`:
```python
if (self.env.options.quit_on_failure
    and pb.instance.status in [TwisterStatus.FAIL, TwisterStatus.ERROR]):
    ...
```